src/mscorlib/shared/System/Memory.cs

```diff
@@ -0,0 +1,287 @@
+
+            if (_index < 0)
+                return new Memory<T>((OwnedMemory<T>)_arrayOrOwnedMemory, (_index & RemoveOwnedFlagBitMask) + start, _length - start);
+            return new Memory<T>((T[])_arrayOrOwnedMemory, _index + start, _length - start);
+        }
+
+        /// <summary>
+        /// Forms a slice out of the given memory, beginning at 'start', of given length
+        /// </summary>
+        /// <param name="start">The index at which to begin this slice.</param>
+        /// <param name="length">The desired length for the slice (exclusive).</param>
+        /// <exception cref="System.ArgumentOutOfRangeException">
+        /// Thrown when the specified <paramref name="start"/> or end index is not in range (&lt;0 or &gt;=Length).
+        /// </exception>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public Memory<T> Slice(int start, int length)
+        {
+            if ((uint)start > (uint)_length || (uint)length > (uint)(_length - start))
+                ThrowHelper.ThrowArgumentOutOfRangeException();
+                
+            if (_index < 0)
+                return new Memory<T>((OwnedMemory<T>)_arrayOrOwnedMemory, (_index & RemoveOwnedFlagBitMask) + start, length);
+            return new Memory<T>((T[])_arrayOrOwnedMemory, _index + start, length);
+        }
+
+        /// <summary>
+        /// Returns a span from the memory.
+        /// </summary>
+        public Span<T> Span
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get
+            {
+                if (_index < 0)
+                    return ((OwnedMemory<T>)_arrayOrOwnedMemory).AsSpan(_index & RemoveOwnedFlagBitMask, _length);
+                return new Span<T>((T[])_arrayOrOwnedMemory, _index, _length);
+            }
+        }
+
+        public unsafe MemoryHandle Retain(bool pin = false)
+        {
+            MemoryHandle memoryHandle;
+            if (pin)
+            {
+                if (_index < 0)
+                {
+                    memoryHandle = ((OwnedMemory<T>)_arrayOrOwnedMemory).Pin();
+                }
+                else
+                {
+                    var handle = GCHandle.Alloc((T[])_arrayOrOwnedMemory, GCHandleType.Pinned);
+                    void* pointer = Unsafe.Add<T>((void*)handle.AddrOfPinnedObject(), _index);
+                    memoryHandle = new MemoryHandle(null, pointer, handle);
+                }
+            }
+            else
+            {
+                if (_index < 0)
+                {
+                    ((OwnedMemory<T>)_arrayOrOwnedMemory).Retain();
+                    memoryHandle = new MemoryHandle((OwnedMemory<T>)_arrayOrOwnedMemory);
+                }
+                else
+                {
+                    var array = (T[])_arrayOrOwnedMemory;
+                    void* pointer = array.Length > 0 ? 
+                                    Unsafe.Add<T>((void*)Unsafe.AsPointer(ref array[0]), _index) :
```

Comment:
@jkotas, instead of using Unsafe.Add, we could do the following, correct? Should we still keep the newly added `Unsafe.Add<T>(void*...)` API?

```C#
void* pointer = array.Length > _index ? (void*)Unsafe.AsPointer(ref array[_index]) : null;
```